### PR TITLE
Fix https://github.com/nrc/rustfmt/issues/377

### PR DESCRIPTION
**Pull request by sinhpham**

### CODE REVIEW

**Comment by marcusklaas** [comment]
Excellent!

The only thing that to consider is when things don't work out. What if there's not enough space, or if a rewrite fails?

We generally propagate such errors to the caller, in the hope that it can create more space for us. Practically, this can be done by returning `None` when a rewrite fails (using `try_opt!`, for example) and checking the width constraints by wrapping the overall result in a `utils::wrap_str` call.

Thanks for the PR!

**Comment by sinhpham** [comment]
Changed it to return None in error cases. I mistakenly thought we need to handle the error at the rewrite func.

File: src/types.rs [hunk 1]
```diff
@@ -482,6 +482,30 @@ impl Rewrite for ast::Ty {
...
+                                                         Some(x)
+                                                     } else {
+                                                         Some(sum + ", " + &x)
+                                                     }
+                                                 }
+                                                 _ => None,
+                                             }
+                                         });
+
+                if let Some(inner) = inner {
```
Comment (marcusklaas) [inline src/types.rs]: Nit: we have the `try_opt!` macro for this. Could you call `wrap_str` on the resulting string? That way we make sure it fits within the given width.

Comment (sinhpham) [inline src/types.rs]: Good point, thanks @marcusklaas

**Comment by eefriedman** [comment]
Maybe consider using write_list instead of just blindly concatenating the members of the tuple?  (The current version isn't wrong, exactly, just less than ideal.)

**Comment by marcusklaas** [comment]
`write_list` doesn't work with separators any more. This is good for now.

Thanks for the PR, @sinhpham!

**Pull request merged**